src/statsmodels/contrasts.jl

```diff
@@ -0,0 +1,233 @@
+    baseind = isnull(C.base) ? 1 : findfirst(c_lvls, convert(eltype(lvls), get(C.base)))
+    baseind > 0 || error("Base level $(C.base) not found in levels")
+    
+    not_base = [1:(baseind-1); (baseind+1):n]
+    tnames = c_lvls[not_base]
+
+    mat = contrasts_matrix(C, baseind, n)
+
+    ContrastsMatrix(mat, tnames, c_lvls, C)
+end
+
+ContrastsMatrix(C::AbstractContrasts, v::PooledDataArray) = ContrastsMatrix(C, levels(v))
+
+
+nullify(x::Nullable) = x
+nullify(x) = Nullable(x)
+
+## Making a contrast type T only requires that there be a method for
+## contrasts_matrix(T, v::PooledDataArray). The rest is boilerplate.
+##
+for contrastType in [:TreatmentContrasts, :SumContrasts, :HelmertContrasts]
+    @eval begin
+        type $contrastType <: AbstractContrasts
+            base::Nullable{Any}
+            levels::Nullable{Vector}
+        end
+        ## constructor with optional keyword arguments, defaulting to Nullables
+        $contrastType(;
+                      base=Nullable{Any}(),
+                      levels=Nullable{Vector}()) = 
+                          $contrastType(nullify(base),
+                                        nullify(levels))
+    end
+end
+
+"""
+    DummyContrasts
+
+One indicator (1 or 0) column for each level, __including__ the base level.
+
+Needed internally when a term is non-redundant with lower-order terms (e.g., in
+~0+x vs. ~1+x, or in the interactions terms in ~1+x+x&y vs. ~1+x+y+x&y. In the
+non-redundant cases, we can (probably) expand x into length(levels(x)) columns
+without creating a non-identifiable model matrix (unless the user has done
+something dumb in specifying the model, which we can't do much about anyway).
+"""
+type DummyContrasts <: AbstractContrasts
+## Dummy contrasts have no base level (since all levels produce a column)
+end
+
+ContrastsMatrix{T}(C::DummyContrasts, lvls::Vector{T}) = ContrastsMatrix(eye(Float64, length(lvls)), lvls, lvls, C)
+
+## Default for promoting contrasts to full rank is to convert to dummy contrasts
+## promote_contrast(C::AbstractContrasts) = DummyContrasts(eye(Float64, length(C.levels)), C.levels, C.levels)
+
+"Promote a contrast to full rank"
+promote_contrast(C::ContrastsMatrix) = ContrastsMatrix(DummyContrasts(), C.levels)
+
+
+
+
+"""
+    TreatmentContrasts
+
+One indicator column (1 or 0) for each non-base level.
+
+The default in R. Columns have non-zero mean and are collinear with an intercept
+column (and lower-order columns for interactions) but are orthogonal to each other.
+"""
+TreatmentContrasts
+
+contrasts_matrix(C::TreatmentContrasts, baseind, n) = eye(n)[:, [1:(baseind-1); (baseind+1):n]]
+
+
+"""
+    SumContrasts
+
+Column for level `x` of column `col` is 1 where `col .== x` and -1 where
+`col .== base`.
+
+Produces mean-0 (centered) columns _only_ when all levels are equally frequent.
+But with more than two levels, the generated columns are guaranteed to be non-
+orthogonal (so beware of collinearity).
```

Comment:
Not sure whether "so beware of collinearity" is really useful here.

Also, in addition to technical details, maybe start with a simpler description? E.g. "each coefficient compares the corresponding level to the average of the other levels" (for treatment, it would be "compares to the baseline", and for Helmert it would be "compares to the average of the preceding levels").